samza-api/src/main/java/org/apache/samza/context/ApplicationDefinedTaskContextFactory.java

```diff
@@ -0,0 +1,20 @@
+package org.apache.samza.context;
+
+import java.io.Serializable;
+
+
+/**
+ * An application should implement this if it has a {@link ApplicationDefinedTaskContext} that is needed for
+ * initialization. This will be used to create instance(s) of that {@link ApplicationDefinedTaskContext}.
+ * <p>
+ * This is {@link Serializable} because it is specified in {@link org.apache.samza.application.ApplicationDescriptor}.
+ * @param <T> concrete type of {@link ApplicationDefinedTaskContext} returned by this factory
+ */
+public interface ApplicationDefinedTaskContextFactory<T extends ApplicationDefinedTaskContext> extends Serializable {
```

Comment:
Just curious. If the user has a custom `ApplicationDefinedTaskContextFactory` or `ApplicationDefinedContainerContextFactory`, how will he provide that to the ApplicationRunner api? Are we planning to take it as a config or a separate high level method in `ApplciationRuntime`(like `register(AppDefinedTaskFactory`))?